File: python/problems/leetcode/0105_construct_binary_tree_from_preorder_and_inorder_traversal/solution.py
"""LeetCode 105: Construct Binary Tree from Preorder and Inorder Traversal."""

from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(
        self, preorder: list[int], inorder: list[int]
    ) -> Optional[TreeNode]:
        idx = {v: i for i, v in enumerate(inorder)}
        it = iter(preorder)

        def build(l: int, r: int) -> Optional[TreeNode]:
            if l > r:
                return None
            v = next(it)
            node = TreeNode(v)
            m = idx[v]
            node.left = build(l, m - 1)
            node.right = build(m + 1, r)
            return node

        return build(0, len(inorder) - 1)


def inorder_traverse(root: Optional[TreeNode]) -> list[int]:
    out: list[int] = []

    def dfs(n: Optional[TreeNode]) -> None:
        if n is None:
            return
        dfs(n.left)
        out.append(n.val)
        dfs(n.right)

    dfs(root)
    return out


if __name__ == "__main__":
    root = Solution().buildTree([3, 9, 20, 15, 7], [9, 3, 15, 20, 7])
    assert inorder_traverse(root) == [9, 3, 15, 20, 7]
    assert Solution().buildTree([], []) is None
    print("0105 OK")

File: cpp/problems/leetcode/0105_construct_binary_tree_from_preorder_and_inorder_traversal/solution.cpp
// LeetCode 105
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for (int i = 0; i < (int)inorder.size(); ++i) idx[inorder[i]] = i;
        pre = &preorder;
        cursor = 0;
        return build(0, (int)inorder.size() - 1);
    }

private:
    unordered_map<int, int> idx;
    vector<int>* pre = nullptr;
    int cursor = 0;
    TreeNode* build(int l, int r) {
        if (l > r) return nullptr;
        int v = (*pre)[cursor++];
        TreeNode* node = new TreeNode(v);
        int m = idx[v];
        node->left = build(l, m - 1);
        node->right = build(m + 1, r);
        return node;
    }
};

static void inorderTraverse(TreeNode* n, vector<int>& out) {
    if (!n) return;
    inorderTraverse(n->left, out);
    out.push_back(n->val);
    inorderTraverse(n->right, out);
}

int main() {
    vector<int> pre{3, 9, 20, 15, 7}, ino{9, 3, 15, 20, 7};
    auto* root = Solution().buildTree(pre, ino);
    vector<int> out;
    inorderTraverse(root, out);
    assert((out == vector<int>{9, 3, 15, 20, 7}));
    vector<int> e1, e2;
    assert(Solution().buildTree(e1, e2) == nullptr);
    cout << "0105 OK" << endl;
    return 0;
}
